nmagent/cachedclient.go

```diff
@@ -0,0 +1,164 @@
+func NewCachedClient(nmagentClient ClientIF) *CachedClient {
+	return &CachedClient{
+		ClientIF: nmagentClient,
+		closing:  make(chan struct{}),
+		// the purpose of the default error here is for handling the case when cns started and already taken request to get HomeAz,
+		// but the initial request to nmagent that sent from the refresh thread is still in process. In this case, the cache value
+		// not get updated in time, so it will return empty homeAzResponse and nil err back, which is misleading. Adding a default
+		// error cache here to resolve that
+		homeAzResponseErrorCache: errors.New("default error cache"),
+	}
+}
+
+// SupportedAPIs retrieves the capabilities of the nmagent running on
+// the node. Cache the response if success
+func (c *CachedClient) SupportedAPIs(ctx context.Context) ([]string, error) {
+	supportedAPIs, err := c.ClientIF.SupportedAPIs(ctx)
+	if err != nil {
+		return supportedAPIs, errors.Wrap(err, "failed to retrieve supportedApis from nmagent")
+	}
+	c.updateNMASupportedApisCache(supportedAPIs)
+	return supportedAPIs, nil
+}
+
+// GetHomeAz returns homeaz cache directly
+func (c *CachedClient) GetHomeAz(_ context.Context) (HomeAzResponse, error) {
+	return c.readHomeAzAndErrorCache()
+}
+
+// isAPISupportedByNMAgent checks cache to see if the given api supported by nmagent client.
+func (c *CachedClient) isAPISupportedByNMAgent(api string) bool {
+	supportedAPIs := c.readNMASupportedApisCache()
+	for _, supportedAPI := range supportedAPIs {
+		if supportedAPI == api {
+			return true
+		}
+	}
+	return false
+}
+
+// readNMASupportedApisCache gets the nmaSupportedApisCache value
+func (c *CachedClient) readNMASupportedApisCache() []string {
+	c.supportedApisMu.RLock()
+	defer c.supportedApisMu.RUnlock()
+	supportedApis := c.nmaSupportedApisCache
+	supportedApisCopy := make([]string, len(supportedApis))
+	copy(supportedApisCopy, supportedApis)
+	return supportedApisCopy
+}
+
+// updateNMASupportedApisCache updates the nmaSupportedApisCache value
+func (c *CachedClient) updateNMASupportedApisCache(supportedApis []string) {
+	c.supportedApisMu.Lock()
+	defer c.supportedApisMu.Unlock()
+	c.nmaSupportedApisCache = supportedApis
+}
+
+// readHomeAzAndErrorCache get the homeAzResponseCache and homeAzResponseErrorCache value
+func (c *CachedClient) readHomeAzAndErrorCache() (HomeAzResponse, error) {
+	c.homeAzMu.RLock()
+	defer c.homeAzMu.RUnlock()
+	homeAzResponse := c.homeAzResponseCache
+	err := c.homeAzResponseErrorCache
+	return homeAzResponse, err
+}
+
+// updateHomeAzAndErrorCache update the homeAzResponseCache and homeAzResponseErrorCache value
+func (c *CachedClient) updateHomeAzAndErrorCache(homeAzResponse HomeAzResponse, err error) {
+	c.homeAzMu.Lock()
+	defer c.homeAzMu.Unlock()
+	c.homeAzResponseCache = homeAzResponse
+	c.homeAzResponseErrorCache = err
+}
+
+// Start starts a new thread to refresh home az cache
+func (c *CachedClient) Start(retryIntervalInSecs int) {
```

Comment:
Rather than take an `int` here, you should take a `time.Duration` to leave units up to the caller. This makes usages self-documenting:

```go
c.Start(2 * time.Second)
```

instead of the cryptic

```go
c.Start(2)
```